h/auth.py

```diff
@@ -0,0 +1,231 @@
+# -*- coding: utf-8 -*-
+"""
+OAuth integration.
+
+Supported grant types
+---------------------
+
+- A client credentials grant using the session and authenticating the client
+  with cross-site request forgery tokens.
+
+Supported token types
+---------------------
+
+- JSON Web Tokens [jwt]_ as bearer tokens.
+
+.. [jwt] https://tools.ietf.org/html/draft-ietf-oauth-json-web-token
+
+Limitations
+-----------
+
+- No support for 3rd party clients exists yet.
+- No support for scopes yet.
+
+"""
+import datetime
+
+import jwt
+from oauthlib.common import generate_client_id
+from oauthlib.oauth2 import RequestValidator as _RequestValidator
+from pyramid.exceptions import BadCSRFToken
+from pyramid import session
+
+from pyramid.authentication import RemoteUserAuthenticationPolicy
+from pyramid.authentication import SessionAuthenticationPolicy
+from pyramid.authorization import ACLAuthorizationPolicy
+from pyramid.util import action_method
+
+from .interfaces import IClientFactory
+
+LEEWAY = 240  # allowance for clock skew in verification
+
+
+class RequestValidator(_RequestValidator):
+
+    """Validates JSON Web Tokens."""
+
+    def authenticate_client(self, request):
+        if request.client_id is None:
+            try:
+                session.check_csrf_token(request, token='assertion')
+            except BadCSRFToken:
+                return False
+            client_id = request.registry.settings['h.client_id']
+            client = get_client(request, client_id)
+            request.client = client
+            request.user = request.session.get('userid')
+        elif request.client_secret is not None:
+            client_id = request.client_id
+            client_secret = request.client_secret
+            client = get_client(request, client_id, client_secret)
+            request.client = client
+            request.user = None
+
+        return request.client is not None
```

Comment:
No. It's set explicitly in both branches.
On Feb 19, 2015 10:48 AM, "Nick Stenning" <notifications@github.com> wrote:

> In h/auth.py
> <https://github.com/hypothesis/h/pull/1951#discussion_r25013170>:
>
> > +            try:
> > +                session.check_csrf_token(request, token='assertion')
> > +            except BadCSRFToken:
> > +                return False
> > +            client_id = request.registry.settings['h.client_id']
> > +            client = get_client(request, client_id)
> > +            request.client = client
> > +            request.user = request.session.get('userid')
> > +        elif request.client_secret is not None:
> > +            client_id = request.client_id
> > +            client_secret = request.client_secret
> > +            client = get_client(request, client_id, client_secret)
> > +            request.client = client
> > +            request.user = None
> > +
> > +        return request.client is not None
>
> Doesn't this imply that if request.client is already set (perhaps by
> accident) this would return True? If that's intended behaviour then this
> is a bit too clever for me. I'd much rather see an early return at the top
> of the function.
>
> If it's *not* intended behaviour, then I'd rather us checking a local
> variable: client is not None.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/hypothesis/h/pull/1951/files#r25013170>.
>